### PR TITLE
feat: lab3 Deploying free5GC

### PR DESCRIPTION
**Pull request by a3828162**

### CODE REVIEW

File: lab3/README.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+  privnet:
+    ipam:
+      driver: default
+      config:
+        - subnet: 10.100.200.0/24
+    driver_opts:
+      com.docker.network.bridge.name: br-free5gc
+```
+For example, privnet is the bridge network for NFs internal communicaion. ex: Nnrf, Nudm...
+
```
Comment (ianchen0119) [inline lab3/README.md:58]: Explain how the bridge network runs, and give an example (e.g. how to use brctl).

File: lab3/exercise/deploy_exercise.yaml [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+      - free5gc-upf
+
+networks:
+  privnet:
+    ipam:
+      driver: default
+      config:
+        - subnet: 10.100.200.0/24
+    driver_opts:
+      com.docker.network.bridge.name: br-free5gc
```
Comment (ianchen0119) [inline lab3/exercise/deploy_exercise.yaml:246]: Add n2, n3, n4, n6 network here.

Comment (ianchen0119) [reply lab3/exercise/deploy_exercise.yaml:246]: Also, describe how to modify the yam file to expose the N2 and N3 interface.

File: lab3/ans/deploy_ans.yaml [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+      - NET_ADMIN
+    networks:
+      n3net:
+        aliases:
+          - upf.n3.org
+        ipv4_address: 10.100.3.100
+      n4net:
+        aliases:
+          - upf.n4.org
+      n6net:
```
Comment (ianchen0119) [inline lab3/ans/deploy_ans.yaml:21]: specify the IP for n6net here.

File: lab3/ans/deploy_ans.yaml [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+
+
```
Comment (ianchen0119) [inline lab3/ans/deploy_ans.yaml:23]: Add n6gw service to forward the traffic from UPF N6 to internet.

File: lab3/ans/deploy_ans.yaml [hunk 5]
```diff
@@ -0,0 +1,303 @@
+version: "3.8"
+
+services:
+  free5gc-upf:
+    container_name: upf
+    image: free5gc/upf:v3.4.2
+    command: bash -c "./upf-iptables.sh && ./upf -c ./config/upfcfg.yaml"
```
Comment (ianchen0119) [inline lab3/ans/deploy_ans.yaml:7]: ip r add default dev `{N6_IF_NAME}` via `{N6_GW_IP}`

**Review by ianchen0119 — CHANGES_REQUESTED** [review]

**Comment by ianchen0119** [comment]
Hi @andy89923 
Please help to review this PR.

Thanks!

File: lab3/ans/deploy_ans.yaml (hunk 5 above)
Comment (a3828162) [reply lab3/ans/deploy_ans.yaml:7]: I write it in **upf-iptable.sh**

File: lab3/ans/deploy_ans.yaml (hunk 4 above)
Comment (ianchen0119) [reply lab3/ans/deploy_ans.yaml:24]: Add additional routing rule in n6 gateway container:
```
ip r add {UE_POOL_CIDR} via {UPF_N6_IP}
```
to forward the UE downlink traffic from Data network to UPF.

File: lab3/README.md [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+Each NF, except for the `UPF`, is assigned an IP address within a `privnet` for internal communication. And assign it an alias for ease of use.
+```yaml
+networks:
+    privnet:
+    aliases:
+        - udr.free5gc.org
+```
+We create three bridge networks named `n1net`, `n2net`, `n3net` and `n6net`. Our goal is respectively assigned network to `AMF`, `SMF`, `UPF` and `UERANSIM`.
```
Comment (ianchen0119) [inline lab3/README.md:89]: `n1net` ?

Comment (ianchen0119) [reply lab3/README.md:89]: All of N1 msgs are carried on NGAP interface.

**Review by andy89923 — COMMENTED** [review]
Suggest cjange ```update here``` to ```TODO```

File: lab3/README.md [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+networks:
+  privnet:
+    ipam:
+      driver: default
+      config:
+        - subnet: 10.100.200.0/24
+    driver_opts:
+      com.docker.network.bridge.name: br-free5gc
+```
+For example, privnet is the bridge network for NFs internal communicaion. ex: Nnrf, Nudm...
```
Comment (andy89923) [inline lab3/README.md:80]: `privnet`
```suggestion
For example, `privnet` is the bridge network for NFs internal communicaion. ex: Nnrf, Nudm...
```

File: lab3/README.md [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+
+Each NF, except for the `UPF`, is assigned an IP address within a `privnet` for internal communication. And assign it an alias for ease of use.
+```yaml
+networks:
+    privnet:
+    aliases:
+        - udr.free5gc.org
```
Comment (andy89923) [inline lab3/README.md:87]:
```suggestion
networks:
    privnet:
      aliases:
        - udr.free5gc.org
```

File: lab3/README.md [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+In `gnbcfg.yaml`, `amfcfg.yaml`, `smfcfg.yaml` and `upfcfg.yaml`, you can find the settings to configure IP addresses as follows.
+```yaml
+  pfcp: # the IP address of N4 interface on this SMF (PFCP)
+    # addr config is deprecated in smf config v1.0.3, please use the following config
+    nodeID: update here # the Node ID of this SMF
+    listenAddr: update here # the IP/FQDN of N4 interface on this SMF (PFCP)
+    externalAddr: update here # the IP/FQDN of N4 interface on this SMF (PFCP)
+```
+Please replace `update here` with the configured N2, N3, and N4 addresses.
```
Comment (andy89923) [inline lab3/README.md:137]: **TODO**

File: lab3/README.md [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+
+Tips: 
+In `smfcfg.yaml`, you will configure the `UPF` N3 interface address because it is required for setting up sessions during SM context creation. If you only use an alias when configuring this address, it may cause DNS resolution issues. Therefore, in `deploy_exercise.yaml`, you should set a static IP address for the `UPF` N3 network and use it here.
+
+After configuring, clone [free5gc-compose](https://github.com/free5gc/free5gc-compose). Then move `free5GLab/lab3/exercise/deploy_exercise.yaml` to `free5gc-compose/` and copy the contents of the files from the `free5GLab/lab3/exercise/config` directory to `free5gc-compose/config`.
+
+You can use these commands to start or stop docker compose.
+```sh
+ce ~/free5gc-compose
```
Comment (andy89923) [inline lab3/README.md:146]: cd?

File: lab3/README.md [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+
+// remove
+docker compose -f deploy_exercise.yaml down
+```
+
+Refer [Create Subscriber via Webconsole](https://free5gc.org/guide/Webconsole/Create-Subscriber-via-webconsole/#5-add-new-subscriber) to create subscriber 
+
+And, attach to ueransim container and run ue
+```sh
+// attach to container
```
Comment (andy89923) [inline lab3/README.md:159]:
```suggestion
// attach to UERANSIM container
```

File: lab3/README.md [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+docker exec -it ueransim bash
+
+// run ue
+./nr-ue -c config/uecfg.yaml
+```
+And the expected result looks like:
+```sh
```
Comment (andy89923) [inline lab3/README.md:166]: I think you forgot to create a subscriber.
However, you could add links to document in [free5gc.org](https://free5gc.org/)

File: lab3/README.md [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+[2024-07-05 12:29:36.962] [nas] [info] Initial Registration is successful
+[2024-07-05 12:29:36.963] [nas] [debug] Sending PDU Session Establishment Request
+[2024-07-05 12:29:36.964] [nas] [debug] UAC access attempt is allowed for identity[0], category[MO_sig]
+[2024-07-05 12:29:36.964] [nas] [debug] Sending PDU Session Establishment Request
+[2024-07-05 12:29:36.964] [nas] [debug] UAC access attempt is allowed for identity[0], category[MO_sig]
+[2024-07-05 12:29:37.173] [nas] [debug] Configuration Update Command received
+[2024-07-05 12:29:37.645] [nas] [debug] PDU Session Establishment Accept received
+[2024-07-05 12:29:37.645] [nas] [info] PDU Session establishment is successful PSI[1]
+[2024-07-05 12:29:37.688] [nas] [debug] PDU Session Establishment Accept received
+[2024-07-05 12:29:37.691] [nas] [info] PDU Session establishment is successful PSI[2]
```
Comment (andy89923) [inline lab3/README.md:198]: I suggest putting a screenshot, and must include the IP address allocation message.

File: lab3/images/architecture.png
Comment (andy89923) [inline lab3/images/architecture.png:1]: N6GW

File: lab3/ans/config/upf-iptables.sh [hunk 14]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+#
+# Configure iptables in UPF
+#
+ip route delete default via 10.100.3.1
+ip route add default via 10.100.6.101 dev dn0
+
+iptables -t nat -A POSTROUTING -o dn0  -j MASQUERADE
```
Comment (andy89923) [inline lab3/ans/config/upf-iptables.sh:8]: NAT here?

File: lab3/README.md [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+* N2: handles control signaling between the `gNB` and `AMF`
+* N3: manages user data transmission between the `gNB` and `UPF`
+* N4: controls user plane configuration and session management between the `SMF` and `UPF`
+
+These three interfaces are the most important interfaces in the 5G system. Therefore, this lab will teach you how to configure the addresses for these interfaces.
+
+### N6 NAT
+Interface to data network.
+
+UPF will performs NAT on packets output through the N6 interface. Rules are set up in `upf-iptables.sh`.
```
Comment (andy89923) [inline lab3/README.md:36]: Is N6GW do the NAT?

Comment (a3828162) [reply lab3/README.md:36]: Yes

File: lab3/ans/config/upf-iptables.sh (hunk 14 above)
Comment (a3828162) [reply lab3/ans/config/upf-iptables.sh:8]: forget remove!

File: lab3/ans/config/uecfg.yaml
Comment (andy89923) [inline lab3/ans/config/uecfg.yaml:1]: I think this file can be delete

File: lab3/ans/config/smfcfg.yaml [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+  userplaneInformation: # list of userplane information
+    upNodes: # information of userplane node (AN or UPF)
+      UPF: # the name of the node
+        type: UPF # the type of the node (AN or UPF)
+        nodeID: upf.n4.org # the Node ID of this UPF
+        addr: upf.n4.org # the IP/FQDN of N4 interface on this UPF (PFCP)
+        interfaces: # Interface list for this UPF
+          - interfaceType: N3 # the type of the interface (N3 or N9)
+            endpoints: # the IP address of this N3/N9 interface on this UPF
+              - 10.100.3.100
```
Comment (andy89923) [inline lab3/ans/config/smfcfg.yaml:16]: Can we write `upf.n3.org`?

File: lab3/ans/config/uecfg.yaml
Comment (a3828162) [reply lab3/ans/config/uecfg.yaml:1]: But I have modified this file, I remove **gnb.free5gc.org**

File: lab3/ans/config/smfcfg.yaml (hunk 16 above)
Comment (a3828162) [reply lab3/ans/config/smfcfg.yaml:16]: SMF is not in N3 net. It cannot resolve this alias. So, I use static ip address.

File: lab3/ans/config/uecfg.yaml
Comment (andy89923) [reply lab3/ans/config/uecfg.yaml:1]: Is this file the same as in `exercise/config/uecfg.yaml`?

Comment (a3828162) [reply lab3/ans/config/uecfg.yaml:1]: Yeah, I think we can just keep one

Comment (andy89923) [reply lab3/ans/config/uecfg.yaml:1]: How about we just point out the changes in `README.md` without the entire file?

Comment (a3828162) [reply lab3/ans/config/uecfg.yaml:1]: I think that's a great idea!

Comment (a3828162) [reply lab3/ans/config/uecfg.yaml:1]: I fix it!

**Review by ianchen0119 — APPROVED** [review]
LGTM

**Pull request merged**